### PR TITLE
perl: Patch config files for missing xlocale.h

### PR DESCRIPTION
**Pull request by iMichka**

### CODE REVIEW

**Comment by iMichka** [comment]
This also fixes the following steps of the postinstall block. All the perl tests seem to be fine after this change. I am not completely sure about the necessity to change the `Config_heavy.pl` file but it make the change more consistent.

**Comment by iMichka** [comment]
Just to be a little bit more clear: once the define is commented out, the xlocale.h is not included in perl.h.

**Comment by maxim-belkin** [comment]
I would suggest going a different route and `s.gsub!` `xlocale.h` to `locale.h`.

**Comment by maxim-belkin** [comment]
note: `locale.h` includes `xlocale.h`, and I believe this steps is independent of `glibc` version.

**Comment by maxim-belkin** [comment]
> iMichka force-pushed the iMichka:perl branch from 72c4d8d to a49d3f5 3 hours ago

💪

**Comment by iMichka** [comment]
In newer version of glibc `xlocale` is completely gone.

perl.h has:
```
#ifdef I_LOCALE
#   include <locale.h>
#endif

#ifdef I_XLOCALE
#   include <xlocale.h>
#endif
```

I_LOCALE is defined, and will work with old an new glibc (because old glibc has it too).

**Comment by maxim-belkin** [comment]
What I meant was that it suffices to replace `xlocale.h` with `locale.h` because `locale.h` includes `xlocale.h` if the latter one exists. At least on CentOS 6 (with glibc 2.6) both are part of glibc-headers package

**Comment by sjackman** [comment]
Is there an upstream bug report for this issue affecting `perl`, and is it resolved or will it be resolved with a newer version of `perl`?

File: Formula/perl.rb [hunk 1]
```diff
@@ -75,10 +77,20 @@ def install
...
-    # CPAN modules installed via the system package manager will not be visible to
-    # brewed Perl. As a temporary measure, install critical CPAN modules to ensure
-    # they are available. See https://github.com/Linuxbrew/homebrew-core/pull/1064
     unless OS.mac?
+      # Glibc does not provide the xlocale.h file since version 2.26
+      # Patch the config files if needed
+      if OS::Linux::Glibc.system_version >= Version.new("2.26")
+        config = "#{prefix}/lib/perl5/#{version}/x86_64-linux-thread-multi/Config_heavy.pl"
+        if File.readlines(config.to_s).grep(/i_xlocale='define'/).size.positive?
+          inreplace config, "i_xlocale='define'", "i_xlocale='undef'"
```
Comment (sjackman) [inline Formula/perl.rb]: You can use `inreplace …, …, …, audit_result: false` to avoid the error when the pattern does not occur in the file.

**Comment by iMichka** [comment]
I could not find any upstream bug for that. Only on package manager websites:
https://bugs.archlinux.org/task/55217
https://bugzilla.redhat.com/show_bug.cgi?id=1464244

They all went for a full Perl rebuild, which is probably the best solution. But we can not afford to do that because we want to use an older Perl build in a portable way. I do not think upstream Perl would be interested to handle this differently: for them everything is fine if you rebuild Perl from source on the target platform. But I may be wrong. Still, even if they want to fix it, I am not sure their config system would allow to dynamically decide what headers to include based on the target platform.

Should we go for the solution from @maxim-belkin ? This would allow to remove the glibc check.

**Comment by sjackman** [comment]
Maxim's fix works for me.

**Comment by iMichka** [comment]
Ready for a second review. CI errors are unrelated.

**Pull request merged**